### PR TITLE
[Data] Fix data test which unconditionally passes

### LINKED ISSUE #41804: [Data] Bad assert on streaming executor tests
In the course of adding [pre-commit hooks](https://github.com/ray-project/ray/pull/41500), two tests got flagged for calling assert on mock methods:

```
check for not-real mock methods..........................................Failed
- hook id: python-check-mock-methods
- exit code: 1

python/ray/data/tests/test_streaming_executor.py:260:    assert o2.add_input.called_once_with(ref1)
python/ray/data/tests/test_streaming_executor.py:264:    assert o2.add_input.called_once_with(ref2)
```

These should be changed to use `assert_called_once_with`; at the moment these will pass unconditionally (!!).

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR fixes a test that unconditionally passes due to improper use of `assert mock.called_once_with`, which was caught by pre-commit checks in #41500.

## Related issue number

Closes https://github.com/ray-project/ray/issues/41804.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by bveeramani — APPROVED** [review]
@peytondmurray for my own understanding, what does `o2.add_input.called_once_with(ref2)` return? Why does it unconditionally pass?

**Comment by peytondmurray** [comment]
@bveeramani `o2.add_input` is a `MagicMock` instance. If you access any attribute or call any 'method' on a `MagicMock` what you get is another `MagicMock` instance - and since `bool(MagicMock()) == True` this test always passes. You could call _any_ attribute of the object we are testing above and the test would pass:

image

So the test isn't actually doing anything at all. On the other hand [`assert_called_with`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with) is a _special_ attribute of `Mock` objects which actually does the test we want.

**Comment by peytondmurray** [comment]
@can-anyscale Can you take a look and merge if this looks okay?

**Comment by can-anyscale** [comment]
Tests are failing now

File: python/ray/data/tests/test_streaming_executor.py [hunk 1]
```diff
@@ -257,11 +257,11 @@ def test_dispatch_next_task():
 
     o2.add_input = MagicMock()
     op_state.dispatch_next_task()
-    assert o2.add_input.called_once_with(ref1)
+    o2.add_input.assert_called_once_with(ref1)
```
Comment (raulchen) [inline python/ray/data/tests/test_streaming_executor.py:260]:
```suggestion
    o2.add_input.assert_called_once_with(ref1, input_index=0)
```
Same for the other one.

**Review by raulchen — APPROVED** [review]

File: python/ray/data/tests/test_streaming_executor.py [hunk 2]
```diff
@@ -257,11 +257,11 @@ def test_dispatch_next_task():
 
     o2.add_input = MagicMock()
     op_state.dispatch_next_task()
-    assert o2.add_input.called_once_with(ref1)
+    o2.add_input.assert_called_once_with(ref1, input_index=0)
 
     o2.add_input = MagicMock()
     op_state.dispatch_next_task()
-    assert o2.add_input.called_once_with(ref2)
+    o2.add_input.assert_called_once_with(ref2, input_index=0)
```
[the remark this answers is not in the document]
Comment (peytondmurray) [reply python/ray/data/tests/test_streaming_executor.py:264]: What pytest version are you using locally? @can-anyscale What pytest version are we using on CI?

[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/data/tests/test_streaming_executor.py:264]: I have 7.0.1

[the remark this answers is not in the document]
Comment (can-anyscale) [reply python/ray/data/tests/test_streaming_executor.py:264]: should line 256 be `op_state.inqueues[0].append(ref2)` for this test to pass?

[the remark this answers is not in the document]
Comment (peytondmurray) [reply python/ray/data/tests/test_streaming_executor.py:264]: @ericl What should this test look like? I'm happy to make the change but I don't have context for what this is supposed to do.

[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/data/tests/test_streaming_executor.py:264]: yes, line 256 should be ref2

**Review by can-anyscale — APPROVED** [review]

**Pull request merged**